File: android/core/data/src/main/kotlin/com/devora/mencare/core/data/fake/FakeAdminRepository.kt
package com.devora.mencare.core.data.fake

import com.devora.mencare.core.common.AppError
import com.devora.mencare.core.common.AppResult
import com.devora.mencare.core.data.repository.AdminRepository
import com.devora.mencare.core.model.AppointmentStatus
import com.devora.mencare.core.model.CampaignSegment
import com.devora.mencare.core.model.DashboardPeriod
import com.devora.mencare.core.model.DashboardStats
import com.devora.mencare.core.model.NotificationSettings
import com.devora.mencare.core.model.OperatorOccupancy
import com.devora.mencare.core.model.PushCampaign
import com.devora.mencare.core.model.UpcomingDay
import com.devora.mencare.core.model.WaitlistStatus
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import java.time.LocalDate
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Dashboard KPIs computed live from the store, on a base offset per period so
 * the demo matches the mockup's order of magnitude ("€ 9.840 · +12%") without
 * seeding hundreds of historical appointments.
 */
private data class PeriodBaseline(
    val revenueCents: Long,
    val appointments: Int,
    val noShows: Int,
    val trendPercent: Int,
)

private fun baselineFor(period: DashboardPeriod): PeriodBaseline = when (period) {
    DashboardPeriod.DAY -> PeriodBaseline(32_000L, 17, 0, 8)
    DashboardPeriod.WEEK -> PeriodBaseline(210_000L, 115, 3, 5)
    DashboardPeriod.MONTH -> PeriodBaseline(900_000L, 490, 14, 12)
}

/** "Prossimi 7 giorni", oggi compreso. */
private const val UPCOMING_DAYS = 7L

@Singleton
class FakeAdminRepository @Inject constructor(
    private val store: InMemoryStore,
) : AdminRepository {

    override fun dashboard(period: DashboardPeriod): Flow<DashboardStats> =
        combine(store.appointments, store.clients, store.operators, store.waitlist) { appointments, clients, operators, waitlist ->
            val today = LocalDate.now()
            val (start, end) = when (period) {
                DashboardPeriod.DAY -> today to today
                DashboardPeriod.WEEK -> today.minusDays((today.dayOfWeek.value - 1).toLong()).let { it to it.plusDays(6) }
                DashboardPeriod.MONTH -> today.withDayOfMonth(1) to today.withDayOfMonth(today.lengthOfMonth())
            }
            val baseline = baselineFor(period)
            val inPeriod = appointments.filter { it.date in start..end }
            val live = inPeriod.filter { it.status != AppointmentStatus.CANCELLED }

            val completed = inPeriod.filter { it.status == AppointmentStatus.COMPLETED }
            val noShows = inPeriod.count { it.status == AppointmentStatus.NO_SHOW } + baseline.noShows
            val revenue = baseline.revenueCents + completed.sumOf { it.totalPriceCents }
            val ticketCount = baseline.appointments + completed.size

            val days = generateSequence(start) { it.plusDays(1) }.takeWhile { it <= end }.toList()
            val perOperator = operators.map { op ->
                val bookedMinutes = live.filter { it.operatorId == op.id }.sumOf { it.durationMinutes }
                val workMinutes = days.sumOf { day ->
                    op.weeklyHours[day.dayOfWeek].orEmpty()
                        .sumOf { java.time.Duration.between(it.start, it.end).toMinutes() }
                }
                OperatorOccupancy(
                    op.id, op.name,
                    if (workMinutes == 0L) 0 else ((bookedMinutes * 100) / workMinutes).toInt().coerceAtMost(100),
                )
            }

            // I prossimi 7 giorni, come li calcola il server: minuti prenotati sui
            // minuti lavorabili di tutti gli operatori, e chi aspetta in lista.
            val upcoming = (0 until UPCOMING_DAYS).map { offset ->
                val day = today.plusDays(offset)
                val salonOpen = store.salon.value.weeklyHours[day.dayOfWeek].orEmpty().isNotEmpty()
                val workable = if (!salonOpen) {
                    0L
                } else {
                    operators.sumOf { op ->
                        op.weeklyHours[day.dayOfWeek].orEmpty()
                            .sumOf { java.time.Duration.between(it.start, it.end).toMinutes() }
                    }
                }
                val booked = appointments
                    .filter {
                        it.date == day && it.status != AppointmentStatus.CANCELLED && it.status != AppointmentStatus.NO_SHOW
                    }
                    .sumOf { it.durationMinutes }
                UpcomingDay(
                    date = day,
                    occupancyPercent = if (workable == 0L) 0 else ((booked * 100) / workable).toInt().coerceAtMost(100),
                    waitlistCount = waitlist.count { it.date == day && it.status == WaitlistStatus.WAITING },
                    closed = workable == 0L,
                )
            }

            DashboardStats(
                period = period,
                revenueCents = revenue,
                revenueTrendPercent = baseline.trendPercent,
                appointmentCount = live.size,
                noShowPercent = if (ticketCount == 0) 0.0 else noShows * 100.0 / ticketCount,
                averageTicketCents = if (ticketCount == 0) 0 else revenue / ticketCount,
                operatorOccupancy = perOperator.sortedByDescending { it.percent },
                upcomingDays = upcoming,
                inactiveClients = clients.filter { it.isInactiveSince(today) },
            )
        }

    override val notificationSettings: Flow<NotificationSettings> = store.notificationSettings

    override suspend fun updateNotificationSettings(settings: NotificationSettings): AppResult<Unit> {
        store.notificationSettings.value = settings
        return AppResult.Success(Unit)
    }

    override val campaigns: Flow<List<PushCampaign>> = store.campaigns

    override suspend fun saveCampaign(campaign: PushCampaign): AppResult<PushCampaign> {
        if (campaign.name.isBlank()) return AppResult.Failure(AppError.Validation("name"))
        if (campaign.body.length > 140) return AppResult.Failure(AppError.Validation("body"))
        val saved = if (campaign.id.isBlank()) campaign.copy(id = store.newId("camp")) else campaign
        val exists = store.campaigns.value.any { it.id == saved.id }
        store.campaigns.value =
            if (exists) store.campaigns.value.map { if (it.id == saved.id) saved else it }
            else store.campaigns.value + saved
        return AppResult.Success(saved)
    }

    override suspend fun reachFor(segment: CampaignSegment): Pair<Int, Int> {
        val today = LocalDate.now()
        val clients = store.clients.value
        val matching = when (segment) {
            CampaignSegment.TUTTI -> clients
            CampaignSegment.INATTIVI_60 -> clients.filter { it.isInactiveSince(today) }
            CampaignSegment.TOP_SPESA -> clients.filter { it.lifetimeSpendCents >= 40_000 }
        }
        // Mockup scale: the fake CRM only holds a handful of records, so pad
        // to the storefront numbers ("198 di 214").
        val size = matching.size + when (segment) {
            CampaignSegment.TUTTI -> 1_272
            CampaignSegment.INATTIVI_60 -> 210
            CampaignSegment.TOP_SPESA -> 80
        }
        val reachable = matching.count { it.marketingOptIn } + when (segment) {
            CampaignSegment.TUTTI -> 1_154
            CampaignSegment.INATTIVI_60 -> 195
            CampaignSegment.TOP_SPESA -> 74
        }
        return reachable to size
    }
}

File: ios/MenCare/Core/Data/Fake/FakeAdminRepository.swift
import Foundation

/// Dashboard KPIs computed live from the store, on a base offset per period so
/// the demo matches the mockup's order of magnitude ("€ 9.840 · +12%") without
/// seeding hundreds of historical appointments.
private struct PeriodBaseline {
    let revenueCents: Int64
    let appointments: Int
    let noShows: Int
    let trendPercent: Int
    /// Divides the monthly synthetic "top services" volumes.
}

private func baselineFor(_ period: DashboardPeriod) -> PeriodBaseline {
    switch period {
    case .day: PeriodBaseline(revenueCents: 32_000, appointments: 17, noShows: 0, trendPercent: 8)
    case .week: PeriodBaseline(revenueCents: 210_000, appointments: 115, noShows: 3, trendPercent: 5)
    case .month: PeriodBaseline(revenueCents: 900_000, appointments: 490, noShows: 14, trendPercent: 12)
    }
}

@MainActor
final class FakeAdminRepository: AdminRepository {

    private let store: InMemoryStore

    init(store: InMemoryStore) {
        self.store = store
    }

    func dashboard(_ period: DashboardPeriod) async throws -> DashboardStats {
        let today = LocalDate.today()
        let start: LocalDate
        let end: LocalDate
        switch period {
        case .day:
            start = today
            end = today
        case .week:
            start = today.minusDays(today.dayOfWeek.rawValue - 1)
            end = start.plusDays(6)
        case .month:
            start = today.startOfMonth
            end = today.endOfMonth
        }
        let baseline = baselineFor(period)
        let inPeriod = store.appointments.filter { $0.date >= start && $0.date <= end }
        let live = inPeriod.filter { $0.status != .cancelled }

        let completed = inPeriod.filter { $0.status == .completed }
        let noShows = inPeriod.filter { $0.status == .noShow }.count + baseline.noShows
        let revenue = baseline.revenueCents + completed.reduce(0) { $0 + $1.totalPriceCents }
        let ticketCount = baseline.appointments + completed.count

        let days = (0...start.daysBetween(end)).map { start.plusDays($0) }
        let perOperator = store.operators.map { op in
            let bookedMinutes = live.filter { $0.operatorId == op.id }.reduce(0) { $0 + $1.durationMinutes }
            let workMinutes = days.reduce(0) { total, day in
                total + (op.weeklyHours[day.dayOfWeek] ?? [])
                    .reduce(0) { $0 + ($1.end.minutesOfDay - $1.start.minutesOfDay) }
            }
            return OperatorOccupancy(
                operatorId: op.id,
                operatorName: op.name,
                percent: workMinutes == 0 ? 0 : min(bookedMinutes * 100 / workMinutes, 100)
            )
        }

        // I prossimi 7 giorni, come li calcola il server: minuti prenotati sui
        // minuti lavorabili di tutti gli operatori, e chi aspetta in lista.
        let upcoming = (0..<7).map { offset -> UpcomingDay in
            let day = today.plusDays(offset)
            let salonOpen = !(store.salon.weeklyHours[day.dayOfWeek] ?? []).isEmpty
            let workable = salonOpen ? store.operators.reduce(0) { total, op in
                total + (op.weeklyHours[day.dayOfWeek] ?? []).reduce(0) { $0 + ($1.end.minutesOfDay - $1.start.minutesOfDay) }
            } : 0
            let booked = store.appointments
                .filter { $0.date == day && $0.status != .cancelled && $0.status != .noShow }
                .reduce(0) { $0 + $1.durationMinutes }
            return UpcomingDay(
                date: day,
                occupancyPercent: workable == 0 ? 0 : min(booked * 100 / workable, 100),
                waitlistCount: store.waitlist.filter { $0.date == day && $0.status == .waiting }.count,
                closed: workable == 0
            )
        }

        return DashboardStats(
            period: period,
            revenueCents: revenue,
            revenueTrendPercent: baseline.trendPercent,
            appointmentCount: live.count,
            noShowPercent: ticketCount == 0 ? 0 : Double(noShows) * 100.0 / Double(ticketCount),
            averageTicketCents: ticketCount == 0 ? 0 : revenue / Int64(ticketCount),
            operatorOccupancy: perOperator.sorted { $0.percent > $1.percent },
            upcomingDays: upcoming,
            inactiveClients: store.clients.filter { $0.isInactiveSince(today) }
        )
    }

    func notificationSettings() async throws -> NotificationSettings { store.notificationSettings }

    func updateNotificationSettings(_ settings: NotificationSettings) async -> AppResult<NotificationSettings> {
        store.notificationSettings = settings
        return .success(settings)
    }

    func campaigns() async throws -> [PushCampaign] { store.campaigns }

    func saveCampaign(_ campaign: PushCampaign) async -> AppResult<PushCampaign> {
        if campaign.name.trimmingCharacters(in: .whitespaces).isEmpty {
            return .failure(.validation(field: "name"))
        }
        if campaign.body.count > 140 { return .failure(.validation(field: "body")) }
        var saved = campaign
        if saved.id.isEmpty {
            saved = PushCampaign(
                id: store.newId("camp"), name: campaign.name, segment: campaign.segment,
                title: campaign.title, body: campaign.body, scheduledAt: campaign.scheduledAt,
                repeatWeekly: campaign.repeatWeekly, sendCap: campaign.sendCap,
                reachableCount: campaign.reachableCount, segmentSize: campaign.segmentSize,
                status: campaign.status
            )
        }
        if store.campaigns.contains(where: { $0.id == saved.id }) {
            let id = saved.id
            let replacement = saved
            store.campaigns = store.campaigns.map { $0.id == id ? replacement : $0 }
        } else {
            store.campaigns.append(saved)
        }
        return .success(saved)
    }

    func reachFor(_ segment: CampaignSegment) async throws -> (reachable: Int, size: Int) {
        let today = LocalDate.today()
        let matching: [ClientRecord]
        switch segment {
        case .tutti: matching = store.clients
        case .inattivi60: matching = store.clients.filter { $0.isInactiveSince(today) }
        case .topSpesa: matching = store.clients.filter { $0.lifetimeSpendCents >= 40_000 }
        }
        // Mockup scale: the fake CRM only holds a handful of records, so pad
        // to the storefront numbers ("198 di 214").
        let sizePad: Int
        let reachablePad: Int
        switch segment {
        case .tutti: sizePad = 1_272; reachablePad = 1_154
        case .inattivi60: sizePad = 210; reachablePad = 195
        case .topSpesa: sizePad = 80; reachablePad = 74
        }
        let reachable = matching.filter(\.marketingOptIn).count + reachablePad
        return (reachable, matching.count + sizePad)
    }
}
